File: libs/datapipe-app/frontend/src/types/opsSpecs.ts
export type OpsColumnKind = "text" | "number" | "datetime" | "duration" | "status" | "link" | "chip" | "split" | "models_count";

export type OpsColumn = {
    id: string;
    label: string;
    source: string;
    kind?: OpsColumnKind;
    fmt?: string | null;
    sortable?: boolean;
    filterable?: boolean;
    link_to?: string | null;
    width?: number | null;
};

export type OpsColumnGroup = {
    label: string;
    columns: OpsColumn[];
};

export type OpsMetricColumn = OpsColumn | OpsColumnGroup;

export type OpsSpecSummary = {
    id: string;
    title: string;
    description: string;
    icon: string;
    color: string;
    has_image: boolean;
    has_model_predictions: boolean;
    has_frozen_datasets: boolean;
    has_training: boolean;
    metric_tables_count: number;
    class_metric_tables_count: number;
    tags?: string[];
};

export type OpsSpecsResponse = {
    pipeline_id: string;
    specs: OpsSpecSummary[];
};

export type OpsTableSchema = {
    id: string;
    title: string;
    table: string;
    metric_source: string;
    primary_columns: OpsColumn[];
    metric_columns: OpsMetricColumn[];
    filters: OpsColumn[];
    default_filters?: OpsFilterRule[];
    default_sort: [string, "asc" | "desc"][];
    entity_links?: Record<string, string>;
};

export type OpsImageOverlayRole = "gt" | "prediction" | "record";

export type OpsImageAnnotationSpecPayload = {
    table: string;
    primary_key_columns: string[];
    bboxes_column?: string | null;
    labels_column?: string | null;
    scores_column?: string | null;
    annotation_column?: string | null;
    join_columns?: Record<string, string>;
    role?: OpsImageOverlayRole;
};

export type OpsImageDataSpecPayload = {
    kind: "image";
    image_table: string;
    image_primary_key_columns: string[];
    image_url_column: string;
    subset_table?: string | null;
    subset_join_columns?: Record<string, string>;
    subset_column?: string | null;
    ground_truth?: OpsImageAnnotationSpecPayload | null;
    visualizer?: string | null;
    preview_size?: number;
    modal_max_side?: number;
    detail_max_side?: number;
};

export type OpsImageRecordViewSpecPayload = {
    kind: "image";
    table: string;
    scope_column?: string | null;
    primary_key_columns: string[];
    image_url_column?: string | null;
    image_url_table?: string | null;
    image_url_join_columns?: Record<string, string>;
    bboxes_column?: string | null;
    labels_column?: string | null;
    scores_column?: string | null;
    width_column?: string | null;
    height_column?: string | null;
    visualizer?: string | null;
    preview_size?: number;
    modal_max_side?: number;
    detail_max_side?: number;
};

export type OpsImagePredictionViewSpecPayload = {
    kind: "image";
    table: string;
    model_id_column: string;
    image_primary_key_columns: string[];
    image_url_table: string;
    image_url_column: string;
    image_url_join_columns?: Record<string, string>;
    prediction?: OpsImageAnnotationSpecPayload | null;
    ground_truth?: OpsImageAnnotationSpecPayload | null;
    subset_table?: string | null;
    subset_join_columns?: Record<string, string>;
    subset_column?: string | null;
    metrics_on_image?: OpsTableSchema | null;
    metrics_on_image_label?: string;
    visualizer?: string | null;
    preview_size?: number;
    modal_max_side?: number;
    detail_max_side?: number;
};

export type OpsTextRecordViewSpecPayload = {
    kind: "text";
    table: string;
    scope_column?: string | null;
    primary_key_columns: string[];
};

export type OpsDataSpecPayload = {
    tables: string[];
    item_table?: string | null;
    label_table?: string | null;
    subset_table?: string | null;
    tag_table?: string | null;
    image_view?: OpsImageDataSpecPayload | null;
};

export type OpsFrozenDatasetSpecPayload = {
    table: string;
    id_column: string;
    created_at_column: string;
    display_name_column?: string | null;
    label_mode?: string;
    split_columns?: Record<string, string>;
    models_count_relation_id?: string | null;
    record_view?: OpsImageRecordViewSpecPayload | OpsTextRecordViewSpecPayload | null;
    columns?: OpsColumn[];
    default_sort?: [string, "asc" | "desc"][];
};

export type OpsModelSpecPayload = {
    table: string;
    id_column: string;
    display_name_column?: string | null;
    created_at_column?: string | null;
    artifact_uri_column?: string | null;
    is_best_table?: string | null;
    is_best_column?: string | null;
    prediction_view?: OpsImagePredictionViewSpecPayload | null;
};

export type OpsTrainingSpecPayload = {
    status_table: string;
    artifact_columns?: Record<string, string>;
    columns?: OpsColumn[];
    default_sort?: [string, "asc" | "desc"][];
};

export type OpsRelationSpecPayload = {
    id: string;
    table: string;
    from_entity: string;
    from_column: string;
    to_entity: string;
    to_column: string;
};

export type OpsSpecDetail = OpsSpecSummary & {
    metrics: OpsTableSchema[];
    class_metrics: OpsTableSchema[];
    data?: OpsDataSpecPayload | null;
    frozen_dataset?: OpsFrozenDatasetSpecPayload | null;
    model?: OpsModelSpecPayload | null;
    training?: OpsTrainingSpecPayload | null;
    relations?: OpsRelationSpecPayload[];
};

export type OpsOverviewResponse = {
    summary: Record<string, unknown>;
    filters?: Record<string, unknown>;
    specs: Array<OpsSpecSummary & Record<string, unknown>>;
    recent_activity?: unknown[];
    lifecycle?: unknown[];
    [key: string]: unknown;
};

export type OpsRowsResponse = {
    rows: Record<string, unknown>[];
    total: number;
    table?: OpsTableSchema;
    columns?: OpsColumn[];
    filter_columns?: OpsColumn[];
    entity_links?: Record<string, string>;
};

export type OpsFilterOperator =
    | "contains"
    | "not_contains"
    | "regex"
    | "equal"
    | "not_equal"
    | "is_empty";

export type OpsFilterMode = "or" | "and";

export type OpsFilterRule = {
    id?: string;
    column_id: string;
    operator: OpsFilterOperator;
    value?: string;
};

export type OpsRowsParams = {
    search?: string;
    sort_by?: string;
    sort_dir?: "asc" | "desc";
    limit?: number;
    offset?: number;
    filter_mode?: OpsFilterMode;
    filters?: OpsFilterRule[];
    /** @deprecated use filters */
    model?: string[] | string;
    /** @deprecated use filters */
    subset?: string[] | string;
};

File: libs/datapipe-ml/datapipe_ml/ops_specs.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal, Sequence

from datapipe_app.specs import (
    OpsSpecBase,
    OpsColumn,
    OpsDataSpec as OpsDataSpecBase,
    OpsMetricTableSpec,
    SnapshotLabelMode,
)
from datapipe_ml.metrics.common import METRICS_NULL_LABEL

ImageOverlayRole = Literal["gt", "prediction", "record"]
RecordViewKind = Literal["image", "text"]


@dataclass(frozen=True)
class OpsImageAnnotationSpec:
    table: str
    primary_key_columns: Sequence[str]
    bboxes_column: str | None = "bboxes"
    labels_column: str | None = "labels"
    scores_column: str | None = None
    annotation_column: str | None = None
    join_columns: dict[str, str] = field(default_factory=dict)
    role: ImageOverlayRole = "record"


@dataclass(frozen=True)
class OpsImageDataSpec:
    kind: Literal["image"] = "image"
    image_table: str = ""
    image_primary_key_columns: Sequence[str] = field(default_factory=tuple)
    image_url_column: str = "image_url"
    subset_table: str | None = None
    subset_join_columns: dict[str, str] = field(default_factory=dict)
    subset_column: str | None = "subset_id"
    ground_truth: OpsImageAnnotationSpec | None = None
    records_show_subset: bool = False
    records_show_ground_truth: bool = False
    visualizer: str | None = None
    preview_size: int = 84
    modal_max_side: int = 1100
    detail_max_side: int = 1600


@dataclass(frozen=True)
class OpsImageRecordViewSpec:
    kind: Literal["image"] = "image"
    table: str = ""
    scope_column: str | None = None
    primary_key_columns: Sequence[str] = field(default_factory=tuple)
    image_url_column: str | None = "image_url"
    image_url_table: str | None = None
    image_url_join_columns: dict[str, str] = field(default_factory=dict)
    bboxes_column: str | None = "bboxes"
    labels_column: str | None = "labels"
    scores_column: str | None = None
    width_column: str | None = None
    height_column: str | None = None
    visualizer: str | None = None
    preview_size: int = 84
    modal_max_side: int = 1100
    detail_max_side: int = 1600


@dataclass(frozen=True)
class OpsImagePredictionViewSpec:
    kind: Literal["image"] = "image"
    table: str = ""
    model_id_column: str = ""
    image_primary_key_columns: Sequence[str] = field(default_factory=tuple)
    image_url_table: str = ""
    image_url_column: str = "image_url"
    image_url_join_columns: dict[str, str] = field(default_factory=dict)
    prediction: OpsImageAnnotationSpec | None = None
    ground_truth: OpsImageAnnotationSpec | None = None
    subset_table: str | None = None
    subset_join_columns: dict[str, str] = field(default_factory=dict)
    subset_column: str | None = "subset_id"
    metrics_on_image: OpsMetricTableSpec | None = None
    metrics_on_image_label: str = METRICS_NULL_LABEL
    visualizer: str | None = None
    preview_size: int = 84
    modal_max_side: int = 1100
    detail_max_side: int = 1600


@dataclass(frozen=True)
class OpsTextRecordViewSpec:
    kind: Literal["text"] = "text"
    table: str = ""
    scope_column: str | None = None
    primary_key_columns: Sequence[str] = field(default_factory=tuple)


@dataclass(frozen=True)
class OpsFrozenDatasetSpec:
    table: str
    id_column: str
    created_at_column: str
    display_name_column: str | None = None
    label_mode: SnapshotLabelMode = "timestamp"
    split_columns: dict[str, str] = field(default_factory=dict)
    models_count_relation_id: str | None = "model_trained_on_frozen_dataset"
    record_view: OpsImageRecordViewSpec | OpsTextRecordViewSpec | None = None
    columns: Sequence[OpsColumn] = field(default_factory=list)
    default_sort: Sequence[tuple[str, Literal["asc", "desc"]]] = field(default_factory=list)


@dataclass(frozen=True)
class OpsModelSpec:
    table: str
    id_column: str
    display_name_column: str | None = None
    created_at_column: str | None = None
    artifact_uri_column: str | None = None
    is_best_table: str | None = None
    is_best_column: str | None = None
    prediction_view: OpsImagePredictionViewSpec | None = None


@dataclass(frozen=True)
class OpsTrainingSpec:
    status_table: str
    artifact_columns: dict[str, str] = field(default_factory=dict)
    columns: Sequence[OpsColumn] = field(default_factory=list)
    default_sort: Sequence[tuple[str, Literal["asc", "desc"]]] = field(default_factory=list)


@dataclass(frozen=True)
class OpsClassMetricTableSpec(OpsMetricTableSpec):
    pass


@dataclass(frozen=True)
class OpsDataSpec(OpsDataSpecBase):
    image_view: OpsImageDataSpec | None = None


@dataclass(frozen=True)
class DatapipeOpsSpec(OpsSpecBase):
    data: OpsDataSpec | None = None
    frozen_dataset: OpsFrozenDatasetSpec | None = None
    model: OpsModelSpec | None = None
    training: OpsTrainingSpec | None = None
    class_metrics: Sequence[OpsClassMetricTableSpec] = field(default_factory=list)


__all__ = [
    "METRICS_NULL_LABEL",
    "DatapipeOpsSpec",
    "OpsClassMetricTableSpec",
    "OpsDataSpec",
    "OpsFrozenDatasetSpec",
    "OpsImageAnnotationSpec",
    "OpsImageDataSpec",
    "OpsImagePredictionViewSpec",
    "OpsImageRecordViewSpec",
    "OpsModelSpec",
    "OpsTextRecordViewSpec",
    "OpsTrainingSpec",
]
